File: 035_Search_Insert_Position/Solution.py
#!python3
# Author: Code-Beast

# MY SOLUTION 1 (Binary search)
# Runtime: 40ms
class Solution:
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int

        >>> Solution().searchInsert([1,3,5,6], 2)
        1
        """
        if len(nums) == 0 or nums[0] >= target:
        	return 0

        if nums[-1] < target:
        	return len(nums)

        left = 0
        right = len(nums) - 1
        while left < right:
        	mid = (left + right) // 2
        	if nums[mid] < target:
        		left = mid + 1
        	elif nums[mid] > target:
        		right = mid - 1
        	else:
        		return mid
        return left if nums[left] >= target else left + 1

File: 035_Search_Insert_Position/Solution.java
// Author: Code-Beast

// MY SOLUTION 1 (Binary search)
// Runtime: 3ms
class Solution {
    public int searchInsert(int[] nums, int target) {
        if (nums.length == 0) return 0;

        int left = 0,
            right = nums.length - 1;
        while (left < right) {
            int mid = (left + right) / 2;
            if (nums[mid] < target) left = mid + 1;
            else if (nums[mid] > target) right = mid - 1;
            else return mid;
        }
        return nums[left] >= target ? left : left + 1;
    }
}
